File: python/tempmail_wrapper/providers/mailnesia.py
"""mailnesia.com provider implementation (HTML scraping, public mailbox)."""

from __future__ import annotations

import random
import re
import string
from datetime import datetime

from bs4 import BeautifulSoup

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class MailnesiaProvider(TempMailProvider):
    """Provider for mailnesia.com — public mailbox, no auth, HTML scraping with IP rotation."""

    BASE_URL = "https://mailnesia.com"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._email: str | None = None
        self._username: str | None = None

    @property
    def name(self) -> str:
        return "mailnesia"

    def _generate_random_ip(self) -> str:
        """Generate a random IP address for header rotation."""
        return f"{random.randint(1, 254)}.{random.randint(0, 255)}.{random.randint(0, 255)}.{random.randint(1, 254)}"

    def _get_headers_with_ip_rotation(self) -> dict[str, str]:
        """Generate headers with rotated IP to bypass 403."""
        ip = self._generate_random_ip()
        return {
            "X-Forwarded-For": ip,
            "X-Real-IP": ip,
            "CF-Connecting-IP": ip,
            "True-Client-IP": ip,
        }

    def generate_email(self) -> str:
        self._username = "".join(
            random.choices(string.ascii_lowercase + string.digits, k=10)
        )
        self._email = f"{self._username}@mailnesia.com"
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        username = email.split("@")[0] if "@" in email else email
        headers = self._get_headers_with_ip_rotation()
        resp = self._http.get(f"{self.BASE_URL}/mailbox/{username}", timeout=30, headers=headers)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to load mailbox: {resp.status_code}", self.name)
        soup = BeautifulSoup(resp.text, "html.parser")
        messages = []
        # Mailnesia uses a table for the inbox
        for row in soup.find_all("tr"):
            cells = row.find_all("td")
            if len(cells) < 3:
                continue
            sender = cells[0].get_text(strip=True)
            subject = cells[1].get_text(strip=True)
            time_str = cells[2].get_text(strip=True)
            # Look for a link with message ID
            link = row.find("a")
            msg_id = ""
            if link:
                href = link.get("href", "")
                msg_id = href.rstrip("/").split("/")[-1]
            try:
                date = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
            except ValueError:
                date = datetime.now()
            if sender or subject:
                messages.append(
                    Message(id=msg_id or subject, sender=sender, subject=subject, date=date)
                )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        if not self._username:
            raise TempMailError("No email generated — call generate_email() first", self.name)
        # Mailnesia message URL: /mailbox/{username}/{message_id}
        url = f"{self.BASE_URL}/mailbox/{self._username}/{message_id}"
        headers = self._get_headers_with_ip_rotation()
        resp = self._http.get(url, timeout=30, headers=headers)
        if resp.status_code != 200:
            raise NotFoundError(f"Message {message_id} not found", self.name)
        soup = BeautifulSoup(resp.text, "html.parser")
        # The message body is typically in a specific div
        body_div = soup.find("div", id="message") or soup.find("div", class_=re.compile(r"message|body|mail"))
        body_html = str(body_div) if body_div else resp.text
        body_text = body_div.get_text(separator="\n", strip=True) if body_div else soup.get_text(strip=True)
        return MessageDetail(
            id=message_id,
            sender="",
            subject="",
            date=datetime.now(),
            body_text=body_text,
            body_html=body_html,
            attachments=[],
        )

    def delete_email(self, email: str) -> bool:
        self._email = None
        self._username = None
        return True

File: rust/src/providers/mailnesia.rs
use async_trait::async_trait;
use regex::Regex;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://mailnesia.com";

/// Provider for mailnesia.com — public mailbox, no auth, HTML table scraping.
#[derive(Debug)]
pub struct Mailnesia {
    client: HttpClient,
    username: Arc<Mutex<Option<String>>>,
}

impl Mailnesia {
    pub fn new(client: HttpClient) -> Self {
        Self {
            client,
            username: Arc::new(Mutex::new(None)),
        }
    }
}

fn random_alphanumeric(len: usize) -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    const CHARSET: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
    let mut s = String::with_capacity(len);
    let mut seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .subsec_nanos() as usize;
    for _ in 0..len {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) % (1 << 31);
        s.push(CHARSET[seed % CHARSET.len()] as char);
    }
    s
}

fn strip_html(html: &str) -> String {
    let re = Regex::new(r"<[^>]+>").unwrap();
    let s = re.replace_all(html, "").replace('\n', " ");
    let ws = Regex::new(r"\s+").unwrap();
    ws.replace_all(&s, " ").trim().to_string()
}

fn parse_date(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S")
        .map(|dt| dt.and_utc())
        .unwrap_or_else(|_| chrono::Utc::now())
}

#[async_trait]
impl TempMailProvider for Mailnesia {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let username = random_alphanumeric(10);
        let email = format!("{}@mailnesia.com", username);
        let mut u = self.username.lock().await;
        *u = Some(username);
        Ok(email)
    }

    async fn get_inbox(&self, email: &str) -> Result<Vec<Message>, TempMailError> {
        let username = if let Some(idx) = email.find('@') {
            &email[..idx]
        } else {
            email
        };
        let headers = self.get_headers_with_ip_rotation();
        let url = format!("{}/mailbox/{}", BASE_URL, username);
        let mut builder = self.client.inner().get(&url);
        for (k, v) in headers {
            builder = builder.header(k, v);
        }
        let resp = builder.send().await.map_err(TempMailError::Http)?;
        if !resp.status().is_success() {
            return Err(TempMailError::Api(format!("HTTP {}", resp.status().as_u16())));
        }
        let html = resp.text().await.map_err(TempMailError::Http)?;
        let re_row = Regex::new(r"(?s)<tr[^>]*>(.*?)</tr>").unwrap();
        let re_cell = Regex::new(r"(?s)<td[^>]*>(.*?)</td>").unwrap();
        let re_link = Regex::new(r#"<a[^>]*href="([^"]*)""#).unwrap();

        let mut messages = Vec::new();
        for row in re_row.captures_iter(&html) {
            let row_html = row.get(1).map(|m| m.as_str()).unwrap_or("");
            let cells: Vec<String> = re_cell
                .captures_iter(row_html)
                .map(|c| strip_html(c.get(1).map(|m| m.as_str()).unwrap_or("")))
                .collect();
            if cells.len() < 3 {
                continue;
            }
            let msg_id = re_link
                .captures(row_html)
                .and_then(|c| c.get(1))
                .map(|m| m.as_str().rsplit('/').next().unwrap_or("").to_string())
                .unwrap_or_default();
            let sender = cells[0].clone();
            let subject = cells[1].clone();
            let time_str = cells[2].clone();
            if sender.is_empty() && subject.is_empty() {
                continue;
            }
            messages.push(Message {
                id: msg_id,
                sender,
                subject,
                date: parse_date(&time_str), preview: String::new(), has_attachments: false });
        }
        Ok(messages)
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let username = self
            .username
            .lock()
            .await
            .clone()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?;
        let html = self
            .client
            .get(&format!("{}/mailbox/{}/{}", BASE_URL, username, message_id))
            .await?;
        let re_msg = Regex::new(r#"(?s)<div[^>]*id="message"[^>]*>(.*?)</div>"#).unwrap();
        let body_html = re_msg
            .captures(&html)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
            .unwrap_or_else(|| html.clone());
        let body_text = strip_html(&body_html);
        Ok(MessageDetail {
            message: Message {
                id: message_id.to_string(),
                sender: String::new(),
                subject: String::new(),
                date: chrono::Utc::now(), preview: String::new(), has_attachments: false },
            body_text: Some(body_text).filter(|s| !s.is_empty()),
            body_html: Some(body_html).filter(|s| !s.is_empty()),
            attachments: vec![], ..Default::default()         })
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let mut u = self.username.lock().await;
        *u = None;
        Ok(true)
    }
}

impl Mailnesia {
    fn generate_random_ip() -> String {
        use std::time::{SystemTime, UNIX_EPOCH};
        let mut seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .subsec_nanos() as u32;
        let mut parts = Vec::with_capacity(4);
        for _ in 0..4 {
            seed = seed.wrapping_mul(1103515245).wrapping_add(12345);
            let val = if parts.is_empty() || parts.len() == 3 {
                (seed % 254) + 1
            } else {
                seed % 256
            };
            parts.push(val.to_string());
        }
        parts.join(".")
    }

    fn get_headers_with_ip_rotation(&self) -> std::collections::HashMap<String, String> {
        let ip = Self::generate_random_ip();
        let mut headers = std::collections::HashMap::new();
        headers.insert("X-Forwarded-For".to_string(), ip.clone());
        headers.insert("X-Real-IP".to_string(), ip.clone());
        headers.insert("CF-Connecting-IP".to_string(), ip.clone());
        headers.insert("True-Client-IP".to_string(), ip);
        headers
    }
}
